internal/ent/hooks/documentdata.go

```
@@ -0,0 +1,129 @@
+	errDocInfoDoesNotMatchAuthenticatedUser = errors.New("NDA submission does not match authenticated user")
+	errUserHasAlreadySignedNDA              = errors.New("user has already signed the NDA")
+	errValidationFailed                     = errors.New("validation failed")
+	errMustBeAnonymousUser                  = errors.New("must be an anonymous user")
+	errMissingResponse                      = errors.New("missing response")
+)
+
+// HookDocumentData runs on document data create mutations to set default values that are not provided
+func HookDocumentData() ent.Hook {
+	return hook.On(func(next ent.Mutator) ent.Mutator {
+		return hook.DocumentDataFunc(func(ctx context.Context, m *generated.DocumentDataMutation) (generated.Value, error) {
+			templateID, _ := m.TemplateID()
+			if templateID == "" {
+				return nil, errMissingTemplate
+			}
+			docTemplate, err := m.Client().Template.Query().Where(template.ID(templateID)).Only(ctx)
+			if err != nil {
+				return nil, err
+			}
+			if docTemplate.Kind != enums.TemplateKindTrustCenterNda {
+				return next.Mutate(ctx, m)
+			}
+
+			anon, ok := auth.AnonymousTrustCenterUserFromContext(ctx)
+			if !ok || anon.SubjectEmail == "" || anon.TrustCenterID == "" || anon.OrganizationID == "" {
+				return nil, errMustBeAnonymousUser
+			}
+
+			response, ok := m.Data()
+			if !ok {
+				return nil, errMissingResponse
+			}
+
+			if err = validateTrustCenterNDAJSON(docTemplate.Jsonconfig, response, anon); err != nil {
+				return nil, err
+			}
+
+			previousDocs, err := m.Client().DocumentData.Query().Where(
+				documentdata.And(
+					documentdata.TemplateIDEQ(docTemplate.ID),
+					func(s *sql.Selector) {
+						s.Where(
+							sqljson.ValueEQ(documentdata.FieldData, anon.SubjectEmail, sqljson.DotPath("signatory_info.email")),
+						)
+					},
+				),
+			).Count(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			if previousDocs > 0 {
+				return nil, errUserHasAlreadySignedNDA
+			}
```

I would check this before the validation of the JSON so you fail fast